src/lowering.rs

```diff
@@ -257,8 +321,27 @@ impl AstVisitorMut for AstLowerer {
     }
 
     fn visit_variable(&mut self, variable: &mut plc_ast::ast::Variable) {
+        let is_recursive = self
+            .ctxt
+            .get_scope()
+            .as_ref()
+            .map(|scope| {
+                self.index.find_pou_type(scope).is_some_and(|it| {
+                    let Ok(pou_type) = self.index.get_effective_type_by_name(it.get_name()) else {
+                        unreachable!("We are in the context of this POU, it must have a type");
+                    };
+                    let var_type = self.index.get_effective_type_or_void_by_name(
+                        variable.data_type_declaration.get_name().unwrap_or_default(),
+                    );
+                    pou_type.get_name() == var_type.get_name()
+                })
+            })
+            .unwrap_or_default();
+        if is_recursive {
+            return;
+        };
         self.maybe_add_global_instance_initializer(variable);
```

Comment:
We could probably do without this if we moved the validation before the lowering step, which I think would be a good idea in general - we do not want diagnostics on internally generated code either way.

Reply:
After trying this out, I've realized that we can no longer call the generated init-functions in user code after this change, since they will be unresolved. I'm not sure how much of a problem this is, it might still be worth it to validate earlier. I'll open an issue for further discussion